electrum/transaction.py

```diff
@@ -1262,6 +412,417 @@ def as_dict(self):
         }
         return out
 
+    @classmethod
+    def estimated_input_weight(cls, txin: dict, is_segwit_tx: bool) -> int:
+        '''Return an estimate of serialized input weight in weight units.'''
+        script = cls.input_script(txin, True)
+        input_size = len(serialize_input(txin, script)) // 2
+
+        if is_segwit_input(txin, guess_for_address=True):
+            witness_size = len(serialize_witness(txin, True)) // 2
+        else:
+            witness_size = 1 if is_segwit_tx else 0
+
+        return 4 * input_size + witness_size
+
+    @classmethod
+    def input_script(cls, txin: dict, estimate_size=False):
+        _type = txin['type']
+        if _type == 'coinbase':
+            return txin['scriptSig']
+
+        # If there is already a saved scriptSig, just return that.
+        # This allows manual creation of txins of any custom type.
+        # However, if the txin is not complete, we might have some garbage
+        # saved from our partial txn ser format, so we re-serialize then.
+        script_sig = txin.get('scriptSig', None)
+        if script_sig is not None and is_txin_complete(txin):
+            return script_sig
+
+        pubkeys, sig_list = get_siglist(txin, estimate_size)
+        script = ''.join(push_script(x) for x in sig_list)
+        if _type == 'address' and estimate_size:
+            _type = guess_txintype_from_address(txin['address'])
+        if _type == 'p2pk':
+            pass
+        elif _type == 'p2sh':
+            # put op_0 before script
+            script = '00' + script
+            redeem_script = multisig_script(pubkeys, txin['num_sig'])
+            script += push_script(redeem_script)
+        elif _type == 'p2pkh':
+            script += push_script(pubkeys[0])
+        elif _type in ['p2wpkh', 'p2wsh']:
+            return ''
+        elif _type == 'p2wpkh-p2sh':
+            pubkey = safe_parse_pubkey(pubkeys[0])
+            scriptSig = bitcoin.p2wpkh_nested_script(pubkey)
+            return push_script(scriptSig)
+        elif _type == 'p2wsh-p2sh':
+            if estimate_size:
+                witness_script = ''
+            else:
+                witness_script = cls.preimage_script(txin)
+            scriptSig = bitcoin.p2wsh_nested_script(witness_script)
+            return push_script(scriptSig)
+        elif _type == 'address':
+            return 'ff00' + push_script(pubkeys[0])  # fd extended pubkey
+        elif _type == 'unknown':
+            return txin['scriptSig']
+        return script
+
+
+class ImmutableTransaction(Transaction):
+    """
+    Used for storing full corresponding tx for selected txin in PSBT input sections
+    """
+
+    def update(self, raw):
+        raise NotImplementedError('Immutable object')
+
+    def input_script(self, txin: dict, estimate_size=False):
+        script_sig = txin.get('scriptSig', '')
+        return script_sig
+
+    def get_witness(self, i: int) -> str:
```

Comment:
this is very confusingly named.
it returns a serialized output (amount + scriptPubKey)...
How about `get_serialized_output` maybe?